File: two_pointers/125.py
# https://leetcode.com/problems/valid-palindrome/
# O(n), O(n)
class Solution:
    def isPalindrome(self, s: str) -> bool:
        l, r = 0, len(s) - 1
        while l < r:
            while l < r and not s[l].isalnum():
                l += 1
            while l < r and not s[r].isalnum():
                r -= 1

            if s[l].lower() != s[r].lower():
                return False

            l += 1
            r -= 1

        return True


if __name__ == "__main__":
    s = "A man, a plan, a canal: Panama"
    print(Solution().isPalindrome(s))

File: two_pointers/125.cpp
#include <iostream>
#include <string>
using namespace std;

// https://leetcode.com/problems/valid-palindrome/
// O(n), O(n)
class Solution {
public:
  bool isPalindrome(string s) {
    int l = 0;
    int r = s.size() - 1;
    while (l < r) {
      while (!isalnum(s[l]) && l < r) {
        l++;
      }
      while (!isalnum(s[r]) && l < r) {
        r--;
      }
      if (tolower(s[l]) != tolower(s[r])) {
        return false;
      }
      l++;
      r--;
    }
    return true;
  }
};

int main(int argc, char *argv[]) {
  string s = "raceacar";
  bool ans = Solution().isPalindrome(s);
  cout << ans << endl;
  return 0;
}
